site-src/geps/gep-726.md

```diff
@@ -0,0 +1,266 @@
+# GEP-726: Add Path Redirects and Rewrites
+
+* Issue: [#726](https://github.com/kubernetes-sigs/gateway-api/issues/726)
+* Status: Implementable
+
+## TLDR
+
+This GEP proposes adding support for path redirects and rewrites in addition to
+host rewrites. This would augment the existing host redirection capabilities.
+
+## Goals
+
+* Implement path redirects.
+* Implement the most portable and simple forms of path rewrites.
+* Describe how more advanced rewrite and redirect and redirect capabilities
+  could be added in the future.
+
+## API
+
+Although many implementations support very advanced rewrite and redirect
+capabilities, the following are the most [portable](#portability) concepts that
+are not already supported by the Gateway API:
+
+* Path redirects
+* Path prefix redirects
+* Path prefix rewrites
+* Host rewrites
+
+Although regular expression based redirects and rewrites are commonly supported,
+there is significantly more variation in both if and how they are implemented.
+Given the wide support for this concept, it is important to design the API in a
+way that would make it easy to add this capability in the future.
+
+### Redirects
+
+The existing `RequestRedirect` filter can be expanded to support path redirects:
+
+```yaml
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: http-filter-1
+spec:
+  rules:
+    - matches:
+      - path:
+          type: Prefix
+          value: /
+      filters:
+      - type: RequestRedirect
+        requestRedirect:
+          hostname: foo.com
+          path:
+            type: Prefix
+            value: /bar
+```
+
+This would be represented with the following API additions:
+```go
+// HTTPRequestRedirect defines a filter that redirects a request. At most one of
+// these filters may be used on a Route rule. This may not be used on the same
+// Route rule as a HTTPRequestRewrite filter.
+type HTTPRequestRedirect struct {
+  // Path defines a path redirect.
+  //
+  // Support: Extended
+  //
+  // +optional
+  Path *HTTPPathRedirect `json:"path,omitempty"`
+  // ...
+}
+
+// HTTPPathRedirectType defines the type of path redirect.
+type HTTPPathRedirectType string
+
+const (
+  // This type of redirect indicates that the complete path will be replaced by
+  // the path redirect value.
+  AbsoluteHTTPPathRedirect HTTPPathRedirectType = "Absolute"
+
+  // This type of redirect indicates that the prefix match will be replaced by
+  // the path redirect value. This type only applies to path matches of type
+  // Prefix. If a Route has other match types, this redirect MUST NOT be applied
+  // to them.
+  PrefixHTTPPathRedirect HTTPPathRedirectType = "Prefix"
```

Comment:
So, for example,

* the request path is `/api/foo/bar
* and the route is a prefix match on `/api`
* and there is a prefix redirect filter with a path of `/overthere`

The final location path would be `/overthere/foo/bar`?

Reply:
Yep, that's exactly the goal, will work on a better explanation for this.

Reply:
Updated to use `ReplacePathPrefixMatch`. Given @jpeach's comment below, I think `ReplacePathMatch` could be a simpler option long term, but we'd need to be very careful when it came to documenting support levels (ie some implementations can only support replacing prefix path matches). In any case, we'll also have to document how this is handled if multiple matches and/or match types are present.

Reply:
OK, I've been thinking about it more, and I think `ReplacePathMatch` is going to provide significantly more flexibility long term. As noted above, it does mean that we'll need to document the subset of path match types that implementations support, but it seems acceptable to continue to consider 100% of this "Extended" support.

Reply:
This has been addressed in more detail in our latest community meeting + https://github.com/kubernetes-sigs/gateway-api/pull/731#issuecomment-964682943 - I ended up going back to `ReplacePathPrefixMatch`.